stream_settings: Fix channel subscription jitter.

adnan-td opened the pull request:
Previously, we were showing loading spinner in each stream's row
on clicking the sub_unsub button which feels like jitter.

Now we are showing plus or checkmark icon directly instead of loading spinner.

Also the `Add` button remains unchanged on clicking subscribe/unsubscribe.

Based on #30557.

[CZO](https://chat.zulip.org/#narrow/stream/9-issues/topic/jitter.20when.20subscribing.2Funsubscribing.20to.20a.20channel/near/1814234)

<details>
<summary>Screenshots and screen captures:</summary>

Before:

![jitter-fix-before](https://github.com/user-attachments/assets/99ec67ca-6a96-44ba-b0e8-d78c73bad012)

After:

![jitter-fix-after](https://github.com/user-attachments/assets/956a15e6-3de6-418f-bbb1-4e4126f6e5b9)
</details>

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

adnan-td committed:
stream_edit: Fix `Add` button enabling on subscribing channel.

In channel settings > subscribers tab, the `Add` button should
stay disabled when add_subscriber input is empty.
`enable_or_disable_add_subscribers_elements` function enables the
input even if it is empty. This feels like jitter.

This PR fixes this behavior by disabling the `Add` button when input is
empty.

shubham-padia commented on web/src/stream_ui_updates.js:
```diff
@@ -448,7 +448,11 @@ export function enable_or_disable_add_subscribers_elements(
+
+        if ($input_element.text().length === 0) {
+            $add_subscribers_button.prop("disabled", true);
+        }
```
It's not clear from the commit message why this change is being made. Please explain the issue the commit is solving in detail in the commit description. Also explain why this change solves that issue and why we would not have regressions after this change.

It would also be good to add a comment here while we're at it.

adnan-td replied:
Added a detailed commit message:

In channel settings > subscribers tab, the `Add` button should
stay disabled when add_subscriber input is empty.
`enable_or_disable_add_subscribers_elements` function enables the
input even if it is empty. This feels like jitter.

This PR fixes this behavior by disabling the `Add` button when input is
empty.

timabbott replied:
From an abstractions standpoint, I feel like `enable_elem` should be being calculated differently, not that we should do something other than using that value. Is there a reason you didn't take that approach?

adnan-td replied:
`enable_or_disable_add_subscribers_elements` enables/disables the whole input:

![image](https://github.com/user-attachments/assets/0bff1d05-3a12-468f-b866-6111b9ac1d2d)

When the input is enabled (usually) (i.e. `enable_elem` is true), the add button is getting enabled even when it is empty. Thus it is being calculated correctly.

The issue becomes clear here:

![add_button_issue](https://github.com/user-attachments/assets/763bf27e-ac97-4ba7-a3de-40b038c7c49b)

The add button gets enabled, when we type something and remove it, it becomes disabled since input is empty.

shubham-padia commented on web/src/stream_settings_components.js:
```diff
@@ -230,39 +180,26 @@ export function unsubscribe_from_private_stream(sub) {
     });
 }
 
-export function sub_or_unsub(sub, $stream_row) {
```
`sub_or_unsub` is still being called with `$stream_row` in other files. Please have a look at those files and remove whatever code is necessary.

adnan-td replied:
Done!

shubham-padia reviewed (changes requested):
Works for me in testing. Left a few comments, rest looks good.

Both the commit messages could benefit from better descriptions which highlights technical decisions + some grammar cleanup.

timabbott commented on web/src/stream_edit.js:
```diff
@@ -596,7 +597,13 @@ export function initialize() {
+            stream_settings_components.sub_or_unsub(sub);
+        } else {
+            stream_settings_components.sub_or_unsub(sub, $stream_row);
+        }
```
Can you explain precisely what happens, ideally with a screenshot where you've added some artificial latency, in the left panel while waiting for the request to return? I'd like to make sure that left-panel state looks good with the spinner removed.

adnan-td replied:
If the `$stream_row` is passed in `sub_or_unsub`, it displays the spinner on that stream in left panel. Here, for the subscribe button, we are not passing it to not show spinner.

<details>
<summary>Screenshots (latency of 3 seconds)</summary>

With spinner (before):

![fix-jitter-loading](https://github.com/user-attachments/assets/960ee6d9-fd42-4f80-bd84-0fbed45caec0)

Spinner removed (after):

![fix-jitter-noloading](https://github.com/user-attachments/assets/f3d07bc4-7346-4385-aea4-36564e2602c3)

</details>

timabbott commented:
@shubham-padia can you review this one again before me?

shubham-padia commented on web/src/stream_edit.js:
```diff
@@ -596,7 +597,13 @@ export function initialize() {
-        stream_settings_components.sub_or_unsub(sub, $stream_row);
+
+        if ($target.hasClass("subscribe-button")) {
+            // Prevents displaying loading spinner in the stream row.
```
It would be good for this comment to specify that we're not showing the spinner when the subscribe/unsubscribe button on the right is clicked. Just reading the if condition does not make it clear.

shubham-padia reviewed (changes requested):
The second commit message hasn't been updated from the previous implementation, it describes the old behaviour. 

Clicking the + button on the left panel still feels jittery to me. 

From https://chat.zulip.org/#narrow/stream/9-issues/topic/.F0.9F.8E.AF.20jitter.20when.20subscribing.2Funsubscribing.20to.20a.20channel/near/1889675, I don't think point no. 3 has been addressed here, that's why the feeling on jitter.

Point 1 and 2 work for me in manual testing and look good codewise.

adnan-td commented:
@shubham-padia addressed the 3rd point mentioned in [CZO](https://chat.zulip.org/#narrow/stream/9-issues/topic/.F0.9F.8E.AF.20jitter.20when.20subscribing.2Funsubscribing.20to.20a.20channel/near/1889675)

Jitter is removed now from left and right sides!

adnan-td committed:
stream_settings: Fix channel subscription jitter.

Fix jitter in left panel of stream settings.

1. Spinner is now triggered only by clicking on the
left-side panel buttons.

2. The spinner has a minimum duration (300ms).

alya commented:
@shubham-padia Please tag for integration review whenever this is ready from your perspective.

shubham-padia commented on web/src/stream_settings_ui.js:
```diff
@@ -67,6 +67,16 @@ export function update_left_panel_row(sub) {
+        setTimeout(() => {
+            stream_settings_components.hide_subscribe_toggle_spinner($new_row);
+        }, 300);
+    }
```
In case of error, the spinner will keep on loading in this case because update_left_panel_row will not be called in case of error, since update_left_panel_row is called when an appropriate event is received. There will be no event in case of error.

Also, I don't like the idea of having the logic to hide the spinner depend on events when it doesn't need to depend on it.

Let's keep the logic for showing and hiding the spinner where the request is being made. 
You can track whether minimum time has elapsed in a variable, and whether the request has completed in another variable. The spinner should be shown when both of them are true. 

I couldn't find a mechanism in loading.js to have minimum time for the loading spinner, but I might be wrong and it's better to check on CZO since it was implied that this mechanism already exists. If not, we still might want to make it part of loading.js, so that it's easier for future contributors to use that mechanism. That conversation needs to happen for this PR.

adnan-td replied:
> Also, I don't like the idea of having the logic to hide the spinner depend on events when it doesn't need to depend on it.

I agree. But if we still require `update_left_panel_row` to wait when the loader is still loading.

In loading.js, we need functions to check in loader is loading, add callback function when loader ends. I've started a discussion in [CZO](https://chat.zulip.org/#narrow/stream/6-frontend/topic/missing.20features.20in.20loading.2Ets/near/1908697) regarding loader.ts

shubham-padia reviewed (changes requested):
Left 1 major comment, the current approach needs some reworking.
